query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/QueryRequestToPinotSQLConverter.java

```diff
@@ -335,39 +433,53 @@ private String getLogicalColumnName(Expression expression) {
     }
   }
 
-  private LiteralConstant[] convertExpressionToMapLiterals(Expression expression) {
+  private LiteralConstant[] convertExpressionToMapLiterals(Expression rhs, Expression lhs) {
     LiteralConstant[] literals = new LiteralConstant[2];
-    if (expression.getValueCase() == LITERAL) {
-      LiteralConstant value = expression.getLiteral();
+    String[] literalArguments = new String[] {"", ""};
+
+    if (rhs.getValueCase() == LITERAL) {
+      LiteralConstant value = rhs.getLiteral();
+      // backward compatibility
       if (value.getValue().getValueType() == ValueType.STRING_ARRAY) {
-        for (int i = 0; i < 2 && i < value.getValue().getStringArrayCount(); i++) {
-          literals[i] =
-              LiteralConstant.newBuilder()
-                  .setValue(
-                      Value.newBuilder().setString(value.getValue().getStringArray(i)).build())
-                  .build();
+        literalArguments[0] = value.getValue().getStringArray(0);
+        if (value.getValue().getStringArrayCount() > 1) {
+          literalArguments[1] = value.getValue().getStringArray(1);
         }
+      } else if (value.getValue().getValueType() == ValueType.STRING) {
+        literalArguments[0] = lhs.getAttributeExpression().getSubpath();
+        literalArguments[1] = value.getValue().getString();
       } else {
         throw new IllegalArgumentException(
             "operator CONTAINS_KEYVALUE supports "
                 + ValueType.STRING_ARRAY.name()
+                + " and "
+                + ValueType.STRING.name()
                 + " value type only");
       }
     }
 
-    for (int i = 0; i < literals.length; i++) {
-      if (literals[i] == null) {
-        literals[i] =
-            LiteralConstant.newBuilder().setValue(Value.newBuilder().setString("").build()).build();
-      }
+    for (int i = 0; i < 2; i++) {
+      literals[i] =
+          LiteralConstant.newBuilder()
+              .setValue(Value.newBuilder().setString(literalArguments[i]).build())
+              .build();
     }
-
     return literals;
   }
 
-  private boolean isComplexAttribute(Expression expression) {
-    return expression.getValueCase().equals(ATTRIBUTE_EXPRESSION)
-        && expression.getAttributeExpression().hasSubpath();
+  private boolean isColumnMapAttribute(Expression expression) {
+    return expression.getValueCase() == COLUMNIDENTIFIER
+        && isMapField(expression.getColumnIdentifier().getColumnName());
+  }
+
+  private boolean isAttributeMapAttribute(Expression expression) {
+    return expression.getValueCase() == ATTRIBUTE_EXPRESSION
+        && expression.getAttributeExpression().hasSubpath()
+        && isMapField(expression.getAttributeExpression().getAttributeId());
```

Comment:
A bad request can also enter. 
eg : 

```
AttributeExpression {
attributeId: API_TRACE.startTime";
subpath: "1234"
}
```

Reply:
If it's an invalid request, we don't want to return false here and just go down some other code path - we want an error. To accomplish that, we should either be doing validation before we even get to this class (rather than inline with transformation), or we just transform it as is and let pinot do the validation for us (this is basically what we do today). The point being, trying to do validation inline with transformation makes for very messy code.

Reply:
I think we can do that once we start deprecating support for columnIdentifier ?

Reply:
resolved